### PR TITLE
Fix tags search and hint while adding tags to test cases in TP

### PR DESCRIPTION
**Pull request by atodorov**
How to reproduce:
- Open a Test Plan
- Go to Cases tab
- Select few test cases
- click Tag -> Add Tag button
- start typing tag name
- if there are tags which match the text they will be shown
  ad a drop-down hint. This was previously raising an exception.

Also simplified the strip_parameters() function which is
used internally and made minor updates to the template and
JavaScript files to match the backend changes and model field names.

Note:
Removal of tags bug is #174 which is fixed in PR #177

### CODE REVIEW

File: tcms/core/ajax.py [hunk 1]
```diff
@@ -137,23 +136,23 @@ def tags(self):
...
+            query = strip_parameters(self.request.GET, self.internal_parameters)
             return User.objects.filter(**query)
 
         def versions(self):
             from tcms.management.models import Version
 
             return Version.objects.filter(product__id=self.product_id)
 
-    objects = Objects(request=request, product_id=request.GET['product_id'])
+    objects = Objects(request=request, product_id=request.GET.get('product_id'))
```
Comment (tkdchen) [inline tcms/core/ajax.py]: Only this line and above https://github.com/Nitrate/Nitrate/pull/178/files#diff-8823511185e05be4d663bf99d5baf21bR129 are required to fix the issue.

Refactor of ``strip_parameters`` should move to a separate patch with tests.

File: tcms/templates/plan/get_cases.html [hunk 2]
```diff
@@ -231,7 +231,7 @@
 		</form>
 	</script>
 	<script id="batch_tag_summary_template" type="text/x-handlebars-template">
-		<div class="dia_title" style=" margin:10px 20px;">You have successfully operate tags in the following case:</div>
+		<div class="dia_title" style=" margin:10px 20px;">You have successfully updated tags for the following case(s):</div>
```
Comment (tkdchen) [inline tcms/templates/plan/get_cases.html]: Please make a separate commit for this fix.

File: tcms/static/js/testplan_actions.js [hunk 3]
```diff
@@ -2468,8 +2468,8 @@ function constructBatchTagProcessDialog(plan_id) {
           'name__startswith': request.term,
           'info_type': 'tags',
           'format': 'ulli',
-          'testcase__plan__pk': plan_id,
-          'field': 'name'
+          'case__plan': plan_id,
+          'field': 'tag__name'
```
Comment (tkdchen) [inline tcms/static/js/testplan_actions.js]: What is this for?

File: tcms/core/ajax.py [hunk 4]
```diff
@@ -137,23 +136,23 @@ def tags(self):
...
     obj = getattr(objects, request.GET['info_type'], None)
 
     if obj:
         if request.GET.get('format') == 'ulli':
             field = request.GET.get('field', 'name')
             response_str = '<ul>'
-            for o in obj():
-                response_str += '<li>' + getattr(o, field, None) + '</li>'
+            for o in obj().values(field):
+                response_str += '<li>' + o.get(field, None) + '</li>'
```
Comment (tkdchen) [inline tcms/core/ajax.py]: Why need to make this change?

File: tcms/core/ajax.py [hunk 5]
```diff
@@ -120,14 +119,14 @@ def env_values(self):
             )
 
         def tags(self):
-            query = strip_parameters(request, self.internal_parameters)
-            tags = TestTag.objects
+            query = strip_parameters(request.GET, self.internal_parameters)
+            tags = TestCaseTag.objects
```
Comment (tkdchen) [inline tcms/core/ajax.py]: Why need to use ``TestCaseTag`` rather than ``TestTag``?

Comment (atodorov) [inline tcms/core/ajax.py]: This is the same reversed child -> parent (or parent -> child) relationship that we talked about in another PR. In this case the relationship goes from TestCaseTag (child) to TestTag (parent). I think it is more natural and more understandable to always use the child -> parent side of the relationship, especially when it is very easy to do so.

The backwards relationship is possible in Django ORM but IMO it is less clear and TBH in many cases it doesn't bring any other benefits. Behind the scenes the parent -> child ORM queries are translated to inner joins which I think are also less efficient.

File: tcms/core/ajax.py (hunk 4 above)
Comment (atodorov) [inline tcms/core/ajax.py]: `obj()` is the `tags()` method from `ajax.py` which returns a query set and then we loop over the objects returned from this query set. Since we only need the value of a single field in this query set `values(field)` makes a small performance optimization and returns the values only for that single field. `obj().values()` in this case is the `QuerySet.values` method.

File: tcms/static/js/testplan_actions.js (hunk 3 above)
Comment (atodorov) [inline tcms/static/js/testplan_actions.js]: This is the result of replacing the TestTag ORM query with TestCaseTag. Using the child class also changes the names of the fields and the AJAX call needs to be updated appropriately.

File: tcms/templates/plan/get_cases.html (hunk 2 above)
Comment (atodorov) [inline tcms/templates/plan/get_cases.html]: Done

File: tcms/core/ajax.py (hunk 1 above)
Comment (atodorov) [inline tcms/core/ajax.py]: added tests

File: tcms/core/ajax.py (hunk 5 above)
Comment (tkdchen) [inline tcms/core/ajax.py]: What do you mean by child -> parent relationship?

Whatever the direction of relationship, here, it only needs to query matched tags by name from ``test_tags`` table (the ``TestTag`` model). In contrast, using ``TestCaseTag`` to query tags requires ``inner join`` which is not good. I don't think this change is necessary.

File: tcms/core/ajax.py (hunk 4 above)
Comment (tkdchen) [inline tcms/core/ajax.py]: Good catch. This should go to a separate patch.

File: tcms/core/ajax.py (hunk 1 above)
Comment (tkdchen) [inline tcms/core/ajax.py]: As mentioned https://github.com/Nitrate/Nitrate/pull/178/commits/8ca779b308b31922bc00fed3b598843e6a78b67f#r116201023, please move other unrelated changes to separate patches.

**Comment by tkdchen** [comment]
ad55cb5 is merged.

I'll merge this 7687d2c after replacing ``REQUEST`` with ``POST`` and ``GET`` properly, because there will be tests.

Please update 8ca779b by removing unrelated code according to my comment. Thanks.

File: tcms/core/tests/test_ajax.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+
+import unittest
+
+from tcms.core.ajax import strip_parameters
+
+
+class Test_strip_parameters(unittest.TestCase):
```
Comment (tkdchen) [inline tcms/core/tests/test_ajax.py]: Please do not use ``_`` in class name. In this case, for example, ``TestStripParameters`` is ok.

**Comment by tkdchen** [comment]
7687d2c is merged.
8ca779b is merged by removing unrelated code according to my comment.
